packages/core/src/components/axes/two-dimensional-hover-axes.ts

```diff
@@ -0,0 +1,122 @@
+			}
+		});
+
+		this.configs.axes = axes;
+
+		// Check the configs to know which axes need to be rendered
+		axisPositions.forEach((axisPositionKey) => {
+			const axisPosition = AxisPositions[axisPositionKey];
+			if (
+				this.configs.axes[axisPosition] &&
+				!this.children[axisPosition]
+			) {
+				const axisComponent = new HoverAxis(this.model, this.services, {
+					position: axisPosition,
+					axes: this.configs.axes,
+					margins: this.margins,
+				});
+
+				// Set model, services & parent for the new axis component
+				axisComponent.setModel(this.model);
+				axisComponent.setServices(this.services);
+				axisComponent.setParent(this.parent);
+
+				this.children[axisPosition] = axisComponent;
+			}
+		});
+
+		Object.keys(this.children).forEach((childKey) => {
+			const child = this.children[childKey];
+			child.render(animate);
+		});
+
+		const margins = {} as any;
+
+		Object.keys(this.children).forEach((childKey) => {
+			const child = this.children[childKey];
+			const axisPosition = child.configs.position;
+
+			// Grab the invisibly rendered axis' width & height, and set margins
+			// Based off of that
+			// We draw the invisible axis because of the async nature of d3 transitions
+			// To be able to tell the final width & height of the axis when initiaing the transition
+			// The invisible axis is updated instantly and without a transition
+			const invisibleAxisRef = child.getInvisibleAxisRef();
+			const {
+				width,
+				height,
+			} = DOMUtils.getSVGElementSize(invisibleAxisRef, { useBBox: true });
+
+			let offset;
+			if (child.getTitleRef().empty()) {
+				offset = 0;
+			} else {
+				offset = DOMUtils.getSVGElementSize(child.getTitleRef(), {
+					useBBox: true,
+				}).height;
+
+				if (
+					axisPosition === AxisPositions.LEFT ||
+					axisPosition === AxisPositions.RIGHT
+				) {
+					offset += 5;
+				}
+			}
+
+			switch (axisPosition) {
+				case AxisPositions.TOP:
+					margins.top = height + offset;
+					break;
+				case AxisPositions.BOTTOM:
+					margins.bottom = height + offset;
+					break;
+				case AxisPositions.LEFT:
+					margins.left = width + offset;
+					break;
+				case AxisPositions.RIGHT:
+					margins.right = width + offset;
+					break;
+			}
+		});
+
+		this.services.events.dispatchEvent(Events.Axis.RENDER_COMPLETE);
+
+		// If the new margins are different than the existing ones
+		const isNotEqual = Object.keys(margins).some((marginKey) => {
+			return this.margins[marginKey] !== margins[marginKey];
+		});
+
+		if (isNotEqual) {
+			this.margins = Object.assign(this.margins, margins);
+
+			// also set new margins to model to allow external components to access
+			this.model.set({ axesMargins: this.margins }, { skipUpdate: true });
+			this.services.events.dispatchEvent(Events.ZoomBar.UPDATE);
+
+			Object.keys(this.children).forEach((childKey) => {
+				const child = this.children[childKey];
+				child.margins = this.margins;
+			});
+
+			this.render(true);
+		}
+	}
+}
```

Comment:
is this an almost identical copy of `TwoDimensionalAxis`?

If yes, then we should avoid that...

For instance, if you'd like to do `new HoverAxis()` you can just set that as an attribute in the class and only extend that...

e.g.

```ts
private AxisComponent = Axis;
```


```ts
private AxisComponent = HoverableAxis;
```

Reply:
Removed `two-dimensional-hover-axes.ts` file. Now cartesian models define what type of axes to create via `protected axisFlavor = AxisFlavor.DEFAULT;` variable. Defined it in main `ChartModelCartesian` set to default and over rid the variable in heatmap model to `AxisFlavor.HOVER`.